pkg/controller/direct/compute/targettcpproxy/targettcpproxy_controller.go

```diff
@@ -38,13 +39,32 @@ import (
 const ctrlName = "firewallpolicyrule-controller"
 
 func init() {
-	registry.RegisterModel(krm.ComputeTargetTCPProxyGVK, NewtargetTCPProxyModel)
+	rg := &TargetTCPProxyReconcileGate{}
+	registry.RegisterModelWithReconcileGate(krm.ComputeTargetTCPProxyGVK, NewTargetTCPProxyModel, rg)
 }
 
-func NewtargetTCPProxyModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+func NewTargetTCPProxyModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
 	return &targetTCPProxyModel{config: config}, nil
 }
 
+type TargetTCPProxyReconcileGate struct {
+	optIn kccpredicate.OptInToDirectReconciliation
+}
+
+var _ kccpredicate.ReconcileGate = &TargetTCPProxyReconcileGate{}
+
+func (r *TargetTCPProxyReconcileGate) ShouldReconcile(o *unstructured.Unstructured) bool {
+	if r.optIn.ShouldReconcile(o) {
+		return true
+	}
+	obj := &krm.ComputeTargetTCPProxy{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(o.Object, &obj); err != nil {
+		return false
+	}
+	// Run the direct reconciler only when spec.location is specified and not global
+	return obj.Spec.Location != nil && obj.Spec.Location != direct.PtrTo("global")
```

Comment:
Does `obj.spec.location` optional and a empty field means it is global? if so, I see a potential user problem. 😢

Reply:
Existing customers(global only) do not need to specify this field, so we have to make `obj.spec.location` optional, to avoid breaking them. Do you have an example for the user problem? 

empty field do you mean `obj.spec.location: ""`? then it will be handled by regional API and the error message looks like: googleapi: Error 400: Invalid value for field 'region': ''. Must be a match of regex '[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?'"